atomics/include/desul/atomics/Compare_Exchange_HIP.hpp

```diff
@@ -17,130 +18,72 @@ SPDX-License-Identifier: (BSD-3-Clause)
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
-  return return_val;
-}
-
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T>
-device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderAcquire, MemoryScope) {
-  atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  return return_val;
-}
-
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4 || sizeof(T) == 8, T>
+template <class T, class MemoryOrder, class MemoryScope>
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T>
 device_atomic_compare_exchange(
-    T* const dest, T compare, T value, MemoryOrderAcqRel, MemoryScope) {
-  atomic_thread_fence(MemoryOrderAcquire(), MemoryScope());
-  T return_val = atomic_compare_exchange(
-      dest, compare, value, MemoryOrderRelaxed(), MemoryScope());
-  atomic_thread_fence(MemoryOrderRelease(), MemoryScope());
-  return return_val;
+    T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+   (void)__hip_atomic_compare_exchange_strong(dest, &compare, value, 
+                                              HIPMemoryOrder<MemoryOrder>::value,
+                                              HIPMemoryOrder<cmpexch_failure_memory_order<MemoryOrder>>::value,
+                                              HIPMemoryScope<MemoryScope>::value);
+  return compare;
 }
 
 template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 4, T> device_atomic_exchange(
-    T* const dest, T value, MemoryOrderRelaxed, MemoryScope) {
-  static_assert(sizeof(unsigned int) == 4,
-                "this function assumes an unsigned int is 32-bit");
-  unsigned int return_val = atomicExch(reinterpret_cast<unsigned int*>(dest),
-                                       reinterpret_cast<unsigned int&>(value));
-  return reinterpret_cast<T&>(return_val);
-}
-template <class T, class MemoryScope>
-__device__ std::enable_if_t<sizeof(T) == 8, T> device_atomic_exchange(
+__device__ std::enable_if_t<atomic_exchange_available_hip<T>::value, T> device_atomic_exchange(
     T* const dest, T value, MemoryOrderRelaxed, MemoryScope) {
-  static_assert(sizeof(unsigned long long int) == 8,
-                "this function assumes an unsigned long long is 64-bit");
-  unsigned long long int return_val =
-      atomicExch(reinterpret_cast<unsigned long long int*>(dest),
-                 reinterpret_cast<unsigned long long int&>(value));
-  return reinterpret_cast<T&>(return_val);
+  T return_val = __hip_atomic_exchange(dest, value,
+                                       HIPMemoryOrder<MemoryOrderRelaxed>::value,
+                                       HIPMemoryScope<MemoryScope>::value);
+  return return_val;
```

Comment:
```suggestion
  return __hip_atomic_exchange(dest, value,
                               HIPMemoryOrder<MemoryOrderRelaxed>::value,
                               HIPMemoryScope<MemoryScope>::value);
```
(haven't checked the format)